File: Difficulty: Medium/Maximum Reachable Index Difference/maximum-reachable-index-difference.py
class Solution:
    def maxIndexDifference(self, s):
        best = [-1] * 26
        ans = -1

        for i in range(len(s) - 1, -1, -1):
            ch = ord(s[i]) - ord('a')
            farthest = i

            if ch < 25 and best[ch + 1] != -1:
                farthest = best[ch + 1]

            best[ch] = max(best[ch], farthest)

            if ch == 0:
                ans = max(ans, farthest - i)

        return ans

File: Difficulty: Medium/Maximum Reachable Index Difference/maximum-reachable-index-difference.java
class Solution {
    public int maxIndexDifference(String s) {
        int n = s.length();
        int[] best = new int[26];
        Arrays.fill(best, -1);
        int ans = -1;
        for (int i = n - 1; i >= 0; i--) {
            int farthest = i;
            int ch = s.charAt(i) - 'a';

            if (ch < 25 && best[ch + 1] != -1) {
                farthest = best[ch + 1];
            }
            best[ch] = Math.max(best[ch], farthest);
            if (ch == 0) {
                ans = Math.max(ans, farthest - i);
            }
        }
        return ans;
    }
}